[MDB Ignore] Monkey to species, take 2

Tsar-Salat opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Title.

closes #8376 
closes #9120

Recoup of #10156 

Reverts #6694, #8962 as they are now obsolete.

Basically, monkeys are species, with all that entails. We are throwing the shitcode away that needed like 10 different refactors over 3 years to deal with, cause the code is fundamentally garbo. 

Humans and Monkeys are already interconnected, no way out of it, so we might as well make the code sane.

Basically, we can make monkeys functionally identical to humans, but we need several explicit drawbacks.
- Absolutely no consoles.
- Probably cant understand most electronic tools
- Some weapon restrictions or two.
- Not roundstart race

Lmk if theres any others that tickle your noggin that you think would make this change make them "OP"

### TODO

- [x] Either fix teratomas or just remove them. Terrible duplicated shitcode.
- [x] fix clothing layering offsets (its kinda annoying to work with, pursuing TG's filter system if Racc doesnt respond)
- [x] audit monkey dexterity
- [x] audit `monkeyize()` & `humanize()`


Ports:
- https://github.com/tgstation/tgstation/pull/55614
- https://github.com/tgstation/tgstation/pull/56620
- https://github.com/tgstation/tgstation/pull/62127
- https://github.com/tgstation/tgstation/pull/69453
- https://github.com/tgstation/tgstation/pull/70356

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

No more aheals when changing race

Hunger is persistent when changing between monkey and humans

All stat vars are not reset inbetween transformations

woaw, genetics dont SUCK

Monkey organs dont runtime the fuck out when they try and do species stuff (who put HUMAN organs inside CARBONS!!!?!)

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.
-->

## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->
<details>
<summary>Screenshots&Videos</summary>

Roundstart monkey ---> human

https://github.com/user-attachments/assets/4113c747-5336-4309-8dfa-a202c0b70a8c

Roundstart human ---> monkey

https://github.com/user-attachments/assets/ae7dda91-a268-4305-8da3-dca6ffb8c2af



Monkeys can hold and *use* two handed weapons, but they do not have the upper core strength to use both arms as a human would.

https://github.com/user-attachments/assets/42dba7c1-38d7-4a85-9f39-d372dc1f5ab9



</details>

## Changelog
:cl: DrDuckedGoose(species design & appearance) rkz(balancing and mass conversions), qustinnus, coiax, IndieanaJones, itseasytosee
add: monkeys are now species
add: monkey tails are now handled like actual tail organs, rather than just being a sprite. They apply and are subtyped the exact same way catgirl tails are, minus a wag animation.
balance: Monkeys can do a lot more things
code: converted mob ventcrawling into flags/traits
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Tsar-Salat committed:
converted

Tsar-Salat committed:
reverts shitty mutations repath

Tsar-Salat committed:
missing stuffs

Tsar-Salat committed:
ventcrawler 1

Tsar-Salat committed:
515

Tsar-Salat committed:
rkidiot

Tsar-Salat committed:
invisible equipment

Tsar-Salat committed:
fix monkey duplicate eyes

Tsar-Salat commented on code/modules/mob/transform_procs.dm:
To be honest, this probably the only file that needs any major revisions. Namely, adding back teratomas (in a sane manner)

Tsar-Salat committed:
readd teratomas

Tsar-Salat committed:
readd changeling ability

DrDuckedGoose committed:
Icon changes

DrDuckedGoose committed:
Icon changes

DrDuckedGoose committed:
Fixes

DrDuckedGoose committed:
Icons for stinky & Fixes

DrDuckedGoose committed:
whoops

Tsar-Salat committed:
Merge branch 'master' into monke-to-species

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
fix helmet

Tsar-Salat committed:
severedmonkey tail sprite to fix unit test

DrDuckedGoose committed:
Merge branch 'monke-to-species' into pr/10943

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
Merge pull request #56 from DrDuckedGoose/pr/10943

DrDuckedGoose monkey & teratoma visual changes, filters, and resizable clothing

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
https://github.com/tgstation/tgstation/pull/62127

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
restraint fix

Tsar-Salat committed:
https://github.com/tgstation/tgstation/pull/56633

Tsar-Salat committed:
https://github.com/tgstation/tgstation/pull/57368

Tsar-Salat committed:
Merge branch 'master' into monke-to-species

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
Merge remote-tracking branch 'Upstream/master' into monke-to-species

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
fixio

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
fix monkey --> human resizing after transform. Roundstart monkeys fix

Tsar-Salat committed:
dual-wield fix

Tsar-Salat committed:
Merge remote-tracking branch 'Upstream/master' into monke-to-species

Tsar-Salat committed:
Merge remote-tracking branch 'Upstream/master' into monke-to-species

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
guh

Tsar-Salat committed:
REALBASELINE

Tsar-Salat committed:
redo search&replace

Tsar-Salat committed:
deltstation

Tsar-Salat committed:
ups

PowerfulBacon commented on code/game/machinery/_machinery.dm:
```diff
@@ -166,6 +166,14 @@ Class Procs:
 
+/obj/machinery/computer/can_interact(mob/user)
+	var/mob/living/living_user = user
+	if(ismonkey(living_user))
```
Would it be better to define a trait for mobs that specifies a lack of dexterity that is included on the species to make this more generic?

Tsar-Salat replied:
So the thing is, there DOES exist a dexterity proc called IsAdvancedToolUser in addition to standard dexterity.

The thing is, its mostly for simplemobs.

For monkeys specifically, it doesn't really make sense for them not to do very basic tasks like press buttons, use a wrench, or be able to remove an embedded piece of glass from their foot.

There's really only 2 things I want to kneecap monkeys on, using computers and using certain weapons. Any check for what Im trying to do can't really be generic, since its really really specific.

If you think a trait will be advantageous, I will go ahead, but I don't really see the use.

EDIT:
I should add, TG does do what you say, adding IS_ADVANCED_TOOLUSER as a species inherent trait for monkeys. Its my opinion thats its pretty... meh. It arbitrarily limits actions which a humanlike being would be perfectly capable of doing.

PowerfulBacon replied:
Generic code like traits is just always cleaner than having specific /typepath checks, I would always avoid checking specific paths in favour of using traits/components/overriding/anything generic where reasonable

Tsar-Salat replied:
alright, Ill work something up for ashies & monkeys, and add code docs. Should be a decent genericization.

Tsar-Salat replied:
Genericized in 0a30ff7

PowerfulBacon commented on code/modules/mob/living/carbon/human/species_types/monkeys.dm:
```diff
@@ -0,0 +1,227 @@
+			return TRUE
+		///REMINDER TO MYSELF TO CORRECT THESE RAND VALUES LATER
+		victim.apply_damage(rand(1, 3), BRUTE, affecting, armor)
+		victim.visible_message("<span class='danger'>[name] bites [victim]!</span>",
```
Does this need any action?

Tsar-Salat replied:
something zesko missed in their RNG removal prs several years ago.

Either I keep current behavior or I try to fit to their vision.

I'll probably keep it the same

PowerfulBacon commented on tools/Runtime Condenser/Input.txt:
This isn't meant to be changed

Tsar-Salat replied:
bad search & find, will fix

Tsar-Salat replied:
fixed in 0a30ff7

PowerfulBacon commented on tools/Runtime Condenser/Output.txt:
This isn't meant to be changed

Tsar-Salat replied:
bad search & find, will fix

Tsar-Salat replied:
fixed in 0a30ff7

PowerfulBacon commented on code/__HELPERS/mobs.dm:
```diff
@@ -128,7 +130,8 @@
 		"body_model" = gender == MALE ? MALE : gender == FEMALE ? FEMALE : pick(MALE, FEMALE),
-		"psyphoza_cap" = pick(GLOB.psyphoza_cap_list)
+		"psyphoza_cap" = pick(GLOB.psyphoza_cap_list),
+		"tail_monkey" = "None",
```
Does this need to be included in the database/have an SQL update?

Tsar-Salat replied:
Monkeys aren't going to be a roundstart species, so we don't need to hook into players preferences. So no SQL, I think.

I will look into the feature, since now that I think about it, this should just be a regular tail on default_features instead of a snowflake feature

PowerfulBacon replied:
Make sure to setup a local database and test that saving characters actually works. I think character saves will fail if something is present in this list but isn't in the database as the insert query will be malformed.

Tsar-Salat replied:
c29a190 Removed the new entry, shouldnt need to update schema anymore

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
tails

Tsar-Salat committed:
woops

Tsar-Salat committed:
code changes?

Tsar-Salat committed:
declare

DrDuckedGoose committed:
If it has a tail...

DrDuckedGoose committed:
Merge branch 'monke-to-species' of https://github.com/Tsar-Salat/BeeStation-Salatland into pr/10943

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
move monkey tails into humans

Tsar-Salat committed:
Merge branch 'master' into monke-to-species

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
fix

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
deltafix

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Rukofamicom commented on code/__HELPERS/mobs.dm:
```diff
@@ -147,6 +147,7 @@
 		"apid_headstripes" = pick(GLOB.apid_headstripes_list),
 		"body_model" = gender == MALE ? MALE : gender == FEMALE ? FEMALE : pick(MALE, FEMALE),
 		"psyphoza_cap" = pick(GLOB.psyphoza_cap_list),
+		"psyphoza_cap" = pick(GLOB.psyphoza_cap_list),
```
```suggestion
```

Rukofamicom replied:
duplicate of the line above, probably a merge conflict moment.

Rukofamicom reviewed (changes requested):
Just a single line I noticed issue with. Easy fix

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
Update code/__HELPERS/mobs.dm

Co-authored-by: Rukofamicom <oozelingcomplaints@gmail.com>

DrDuckedGoose commented on code/datums/dna.dm:
```diff
@@ -714,3 +709,14 @@
+	//ALL our important visual stuff gets reset when we update appearance, so we have to set it back
+	height_displacement_holder.vis_flags = VIS_UNDERLAY | VIS_INHERIT_DIR
+	height_displacement_holder.appearance_flags = TILE_BOUND | PIXEL_SCALE | RESET_TRANSFORM
+	height_displacement_holder.plane = LOWEST_EVER_PLANE
```
```suggestion
	height_displacement_holder.plane = PLANE_SPACE
	height_displacement_holder.layer = 0
```

DrDuckedGoose committed:
Update code/datums/dna.dm

PowerfulBacon commented on code/datums/dna.dm:
```diff
@@ -714,3 +709,15 @@
+	height_displacement_holder.appearance_flags = TILE_BOUND | PIXEL_SCALE | RESET_TRANSFORM
+	height_displacement_holder.plane = PLANE_SPACE
+	height_displacement_holder.layer = 0
+	height_displacement_holder.render_target = "[ref(height_displacement_holder)]"
```
```suggestion
	height_displacement_holder.render_target = "*[REF(height_displacement_holder)]"
```

* makes it not render to the screen and only draws to itself. Tested in game and this seems to works fine.

![image](https://github.com/user-attachments/assets/d12ff64e-7924-4909-9a04-71710a1b279b)

PowerfulBacon commented on code/datums/dna.dm:
```diff
@@ -18,21 +18,23 @@
 		holder = new_holder
-	height_displacement = icon('icons/effects/64x64.dmi', "height_displacement")
+	update_height_holder()
+	new_holder?.vis_contents += height_displacement_holder
```
To be honest, I'm not really a fan of this change since its a little bit janky and breaks the right click menu. Does the monkey height displacement really have to be directional?

I would like to note that displacement maps do not update the hitbox for a sprite.

The level of jankiness here kind of feels like it offsets the advantages of converting monkeys into a species in the first-place and is something I feel like definitely needs some action to be taken. The most optimal solution would be finding some kind of displacement map that can just work for all directions in a way thats good enough. Shrinking the sprite vertically is always going to make it look a bit ugly just because its quite a big restriction on how monkeys can look. If there is absolutely nothing that can be done then... I guess we kind of have to live with it a bit.

If absolutely nothing can be done, and we have to have a vis_contents displacement map that has to respond to directions then does the map have to be 64x64 instead of 32x32? If it could be 32x32 then the alpha could be set to 0 which would limit its effect on the right click menu.

DrDuckedGoose replied:
>  Does the monkey height displacement really have to be directional?

Yes, it also opens up future stuff for other species. I don't think it's that big of an issue tbh

> then the alpha could be set to 0

Breh

Tsar-Salat replied:
> Yes, it also opens up future stuff for other species.

![image](https://github.com/user-attachments/assets/4459d8c6-a2bc-4bef-b47b-60b86e374b4b)

If we want to do unique stuff with height displacement, I think this is a good idea.

Having an excessively short race other than monkeys would make sense, given... basically every race is within the same margin as regular humans, which is rather anthropocentric... 

See Shiptest's short dinosaurian-bird race for this done fairly well.

DrDuckedGoose replied:
You might even be able to revive Gr*d with this new approach

Tsar-Salat committed:
fixes examine

Tsar-Salat committed:
that shit too

Tsar-Salat committed:
rkstupid

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
Update code/datums/dna.dm

Co-authored-by: PowerfulBacon <26465327+PowerfulBacon@users.noreply.github.com>

DrDuckedGoose committed:
fix popup menu visuals - 25 10 24

Fixes issues with the displacement filter fucking up the right click menu
This solution will probably need a second opinion

DrDuckedGoose committed:
Oops

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

DrDuckedGoose committed:
Is this jank?

Anyone can code, only the best can cook

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

DrDuckedGoose committed:
bump

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

PowerfulBacon reviewed (changes requested):
I want to put approval on this, but this height displacement solution is honestly just... bad. TG has a far better approach, instead of trying to have a single height map for squishing everything, they have a displacement filter for stretching/squishing legs and stretching/squishing the body. They don't need to do any janky code for direction displacement maps at all, which if possible I would much, much prefer if this could use that. It might need some changes to the monkey sprites to accomodate but if they are sprited like humans and squished afterwards (same way TG's are) then it shouldn't cause an issue.

Until byond gets support for directional heightmaps, we should avoid them where possible and TG has shown that it is possible to avoid them.

Worst comes to worst, how do monkeys look using the old method for displacement maps, or a displacement map that doesn't use directional states? I would rather have a single heightmap that looks good enough in most scenarios even if it is imperfect than have a complete mess trying to deal with the directional code.

Tsar-Salat committed:
Merge branch 'master' into monke-to-species

Tsar-Salat committed:
Merge remote-tracking branch 'BeeStation/master' into monke-to-species

Tsar-Salat committed:
baseline

Tsar-Salat committed:
sorry racc

Tsar-Salat committed:
Merge remote-tracking branch 'upstream/master' into monke-to-species

# Conflicts:
#	_maps/RandomRuins/LavaRuins/lavaland_surface_syndicate_base1.dmm
#	_maps/map_files/BoxStation/BoxStation.dmm
#	_maps/map_files/FlandStation/FlandStation.dmm
#	_maps/map_files/RadStation/RadStation.dmm
#	beestation.dme
#	code/__DEFINES/DNA.dm
#	code/_globalvars/traits/_traits.dm
#	code/_onclick/hud/monkey.dm
#	code/_onclick/other_mobs.dm
#	code/datums/components/creamed.dm
#	code/datums/components/twohanded.dm
#	code/datums/dna.dm
#	code/datums/mutations.dm
#	code/datums/mutations/__mutations.dm
#	code/datums/mutations/_combined.dm
#	code/datums/mutations/_mutations.dm
#	code/datums/mutations/actions.dm
#	code/datums/mutations/body.dm
#	code/datums/mutations/chameleon.dm
#	code/datums/mutations/cold.dm
#	code/datums/mutations/hulk.dm
#	code/datums/mutations/radioactive.dm
#	code/datums/mutations/sight.dm
#	code/datums/mutations/speech.dm
#	code/datums/mutations/touch.dm
#	code/game/machinery/_machinery.dm
#	code/game/objects/items.dm
#	code/game/objects/items/devices/scanners/scanners.dm
#	code/game/objects/items/dna_injector.dm
#	code/game/objects/items/stacks/sheets/mineral/glass.dm
#	code/game/objects/structures/petrified_statue.dm
#	code/game/objects/structures/plasticflaps.dm
#	code/game/turfs/turf_integrity.dm
#	code/modules/admin/verbs/randomverbs.dm
#	code/modules/antagonists/abductor/equipment/gland.dm
#	code/modules/antagonists/changeling/changeling_power.dm
#	code/modules/antagonists/changeling/powers/tiny_prick.dm
#	code/modules/antagonists/revenant/revenant_abilities.dm
#	code/modules/atmospherics/machinery/atmosmachinery.dm
#	code/modules/client/verbs/suicide.dm
#	code/modules/jobs/access.dm
#	code/modules/mob/living/carbon/alien/humanoid/queen.dm
#	code/modules/mob/living/carbon/carbon_defense.dm
#	code/modules/mob/living/carbon/human/examine.dm
#	code/modules/mob/living/carbon/human/human_defense.dm
#	code/modules/mob/living/carbon/human/human_defines.dm
#	code/modules/mob/living/carbon/human/monkey.dm
#	code/modules/mob/living/carbon/life.dm
#	code/modules/mob/living/carbon/monkey/inventory.dm
#	code/modules/mob/living/carbon/monkey/life.dm
#	code/modules/mob/living/carbon/monkey/monkey.dm
#	code/modules/mob/living/carbon/monkey/monkey_defense.dm
#	code/modules/mob/living/living.dm
#	code/modules/mob/living/living_defense.dm
#	code/modules/mob/living/simple_animal/friendly/cat.dm
#	code/modules/mob/living/simple_animal/friendly/crab.dm
#	code/modules/mob/living/simple_animal/friendly/farm_animals.dm
#	code/modules/mob/living/simple_animal/friendly/snake.dm
#	code/modules/mob/living/simple_animal/hostile/giant_spider.dm
#	code/modules/mob/living/simple_animal/hostile/killertomato.dm
#	code/modules/mob/living/simple_animal/hostile/mimite.dm
#	code/modules/mob/living/simple_animal/hostile/mining_mobs/gutlunch.dm
#	code/modules/mob/living/simple_animal/hostile/redgrub.dm
#	code/modules/mob/living/simple_animal/hostile/retaliate/frog.dm
#	code/modules/mob/living/ventcrawling.dm
#	code/modules/mob/transform_procs.dm
#	code/modules/projectiles/projectile/magic.dm
#	code/modules/reagents/chemistry/reagents/food_reagents.dm
#	code/modules/research/xenobiology/xenobio_camera.dm
#	code/modules/spells/spell_types/barnyard.dm
#	code/modules/spells/spell_types/godhand.dm
#	code/modules/spells/spell_types/touch_attacks.dm

Tsar-Salat committed:
Revert "sorry racc"

This reverts commit cf25a1855be05b39eac96a018374f29ad6cf7be7.

# Conflicts:
#	code/modules/surgery/bodyparts/bodyparts.dm

Tsar-Salat committed:
REALBASELINE

Tsar-Salat committed:
update paths

Tsar-Salat committed:
Merge remote-tracking branch 'upstream/master' into monke-to-species

Tsar-Salat committed:
discoordinated status effect and brain traits

https://github.com/tgstation/tgstation/pull/65278

Tsar-Salat committed:
combine left and right arms (it was pissing me off)

Tsar-Salat committed:
fat unarmed_attack refactor (blows up paw attacks with warhead)

https://github.com/tgstation/tgstation/pull/78991 https://github.com/tgstation/tgstation/pull/66711 https://github.com/tgstation/tgstation/pull/45866

Tsar-Salat committed:
fixes

Tsar-Salat committed:
Merge remote-tracking branch 'upstream/master' into monke-to-species
